### ISSUE TITLE
API error response doesn't follow JSON RPC spec

### ISSUE THREAD

**Issue by shaneneuerburg** [cgrates/cgrates#565]
I'm using a third-party JSON RPC client that follows the JSON RPC 2.0 specifications. When it checks for the successful state of the response, it expects to see an error object with a `code` property. However, CGRates is only returning the message.

Here is what the error should look like:
http://www.jsonrpc.org/specification#error_object

I'm not sure what the best course of fixing this is, as modifying the behavior of the current API could result in other people's implementations breaking.

**Comment by danbogos** [comment]
Hi Shane

Thanks for your report!

It would be great if you could attach a ngrep trace of your request/reply
as seen over network so I can understand more about the issue.

Thanks
DanB
Am 19.10.2016 22:09 schrieb "Shane Neuerburg" <notifications@github.com>:

**Comment by shaneneuerburg** [comment]
Here is an example request to ApierV1.DebitUsage (I intentionally sent in invalid params to test error conditions):
```
#########
T 10.211.55.1:58061 -> 10.211.55.19:2080 [AP]
  POST /jsonrpc HTTP/1.1..Host: localhost:2080..Content-Length: 228..User-Ag
  ent: JSON-RPC PHP Client <https://github.com/fguillot/JsonRPC>..Content-Typ
  e: application/json..Accept: application/json..Connection: close....{"jsonr
  pc":"2.0","method":"ApierV1.DebitUsage","id":148051265,"params":[{"Tenant":
  "example.com","ToR":"phone_card","Category":"foo","Subject":"1001","Direction"
  :"*out","Account":"100000000","Destination":"1112223333","Usage":"1"}]}
##
T 10.211.55.19:2080 -> 10.211.55.1:58061 [AP]
  HTTP/1.1 200 OK..Content-Type: application/json..Date: Wed, 19 Oct 2016 08:
  51:24 GMT..Content-Length: 77..Connection: close....{"id":148051265,"result
  ":null,"error":"SERVER_ERROR: RATING_PLAN_NOT_FOUND"}.
#####
```

As you can see in the result, you are sending back the error like this:
```json
{
  "id":148051265,
  "result":null,
  "error":"SERVER_ERROR: RATING_PLAN_NOT_FOUND"
}
```

According to the spec, it should return more like this:
```json
{
  "id":148051265,
  "result":null,
  "error": {
     "code":1234, 
     "message":"SERVER_ERROR: RATING_PLAN_NOT_FOUND", 
     "data":""
  }
}
```

**Comment by shaneneuerburg** [comment]
It looks like the all JSON RPC specs since 1.1 have required the error to have the above structure. 1.0 was ambiguous on what it expected other than being null when successful.

**Comment by danbogos** [comment]
Checking it appears you are right. We (and Go standard library) are still
with version 1. For the moment we are only supporting  1 but we could add
another transport and server I guess if in high demand in the future,
however I would prefer that Go changes that first.

You can find an issue opened in Go here:
https://github.com/golang/go/issues/10929

DanB
Am 19.10.2016 22:58 schrieb "Shane Neuerburg" <notifications@github.com>:

**Comment by shaneneuerburg** [comment]
Oh, wow, I didn't realize go still hadn't implemented 2.0. I guess I'll have to work around it somehow. Thanks Dan!

**Comment by cgrates** [comment]
Sure. Honestly I felt many times that error category is something missing
from the language itself.
Enjoy!

DanB
